examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2076 @@
+struct State<S> {
+    /// The state for each address.
+    state:               StateMap<Address, AddressState<S>, S>,
+    /// All of the token IDs
+    all_tokens:          StateSet<ContractTokenId, S>,
+    /// Map with contract addresses providing implementations of additional
+    /// standards.
+    implementors:        StateMap<StandardIdentifierOwned, Vec<ContractAddress>, S>,
+    /// A registry to link an account to an public key and its nonce. The
+    /// corresponding private key registered here has full access to the
+    /// tokens controlled by the account. The nonce is used to prevent replay
+    /// attacks of signed transactions.
+    public_key_registry: StateMap<AccountAddress, (PublicKeyEd25519, u64), S>,
+}
+
+/// Part of the parameter type for the contract function `registerPublicKey`.
+/// Takes the account and the public key that should be linked.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct RegisterPublicKeyParam {
+    /// Account that a public key will be registered to.
+    account:    AccountAddress,
+    /// The public key that should be linked to the above account.
+    public_key: PublicKeyEd25519,
+}
+
+/// The parameter type for the contract function `setImplementors`.
+/// Takes a standard identifier and list of contract addresses providing
+/// implementations of this standard.
+#[derive(Debug, Serialize, SchemaType)]
+struct SetImplementorsParams {
+    /// The identifier for the standard.
+    id:           StandardIdentifierOwned,
+    /// The addresses of the implementors of the standard.
+    implementors: Vec<ContractAddress>,
+}
+
+/// Part of the parameter type for the contract function `permit`.
+/// Specifies the transaction payload that should be forwarded to either the
+/// `transfer` or the `updateOperator` function.
+#[derive(Debug, Serialize, SchemaType, Clone)]
+enum PermitPayload {
+    Transfer(TransferParameter),
+    UpdateOperator(UpdateOperatorParams),
+}
+
+/// Part of the parameter type for the contract function `permit`.
+/// Specifies the message that is signed.
+#[derive(SchemaType, Serialize)]
+struct PermitMessage {
+    /// The contract_address that the signature is intended for.
+    contract_address: ContractAddress,
+    /// The entry_point that the signature is intended for.
+    entry_point:      OwnedEntrypointName,
+    /// A nonce to prevent replay attacks.
+    nonce:            u64,
+    /// A timestamp to make signatures expire.
+    timestamp:        Timestamp,
+    /// The payload that should be forwarded to either the `transfer` or the
+    /// `updateOperator` function.
+    payload:          PermitPayload,
+}
+
+/// The parameter type for the contract function `permit`.
+/// Takes a signature, the signer, and the message that was signed.
+#[derive(Serialize, SchemaType)]
+pub struct PermitParam {
+    /// Signature.
+    signature: SignatureEd25519,
+    /// Signer that signed the above signature.
+    signer:    AccountAddress,
```

Comment:
I think we should support the same type of signature, which is a multisig like we do for transactions.

That will make it simpler to later use accounts. That signature is essnetially just a map
```
BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>
```

Reply:
I implemented the following: The CIS3 standard supports multi-sig accounts but because the `public_key_registry` in this contract maps one public key to one account, only one signature has to be provided for this contract. This signature has to be at the key 0 in both maps below:
```
  signature: BTreeMap<u8, BTreeMap<u8, SignatureEd25519>>
 ```